### DISCUSSION TITLE
Face Swap Feature ?

### DISCUSSION THREAD

**Discussion by drdancm** [lllyasviel/Fooocus#474] [Ideas]
I use Fooocus inside Stability Matrix and it works beautifully. However with the other packages (A111, Invoke, SD.Next) many of us newer users who are pretty much completely naive about python and programming, would love to have that feature because it is so much fun to put your own face on comic strip characters as well as populate them with our friends.

While many advanced users have Roop, it has now been discontinued and my and other people's efforts to install Face Swap Lab, or Gourieff's Reactor have completely failed after much effort and lots of advice. Mostly insightface is not found etc. Apparently those programs always expect the user to know how to use python etc. to install pre-requisites. Usually the advice they give is to install A, B, C  but until a friend of mine helped and explained that some things need to be run from Command (in Windows), other things must be run in a specific version of Python. He had to show me how to quit that program since neither Quit or Exit did it for me. This may be a hilarious to some people familiar with these programs, but it is very frustrating for those us unfamiliar with things like this, who don't know the difference between Python, Torch, Pytorch,  Gradio, or Pip (at this point I finally have _some_ idea ) :

Because you intended Fooocus to be super easy to use (and it really is!) is the reason why I'm asking if you would entertain the idea of making some sort of Face Swap feature for your excellent _easiest-of-all-to use_ program.

Thanks,

Dan

**Comment by barepixels** [comment]
Maybe this will help

**ComfyUI Fundamentals - Face Restoration**
https://youtu.be/FShlpMxbU0E?si=tuIHYCA7PhLN6-cP
Link to workflow json  https://drive.google.com/file/d/1TCa9EdqNiAJ7QFM7rPHY_a1tLexk_8Ol/view

**Reply by KINGLIFER** [reply]
What that has to do with Fooocus?

**Reply by barepixels** [reply]
From what I understand Fooocus tries to be like Midjourney for Stable Diffussion. 
Well [You Can Now Change Faces On An Image with Midjourney!](https://bootcamp.uxdesign.cc/you-can-now-swap-face-instantly-with-midjourney-and-this-bot-84ef028ddcad)

I need to look more into Face Swap Lab.  Apparently, you can build a model of a person face and reuse the model 

I did played with [IPAdapter_face ](https://github.com/MoonRide303/Fooocus-MRE/discussions/164) and was blown away by how well it worked.  See the link

**Comment by drdancm** [comment]
With A1111 I finally succeeded getting Face Swap Lab to work. The key was using a pre-built insight face  to substitute for having to install Microsoft Visual Studio etc.

see here:
https://github.com/Gourieff/sd-webui-reactor#insightfacebuild

** Here is the crucial part that helped get it working**
_[VIII. (For Windows users) If you still cannot build Insightface for some reasons or just don't want to install Visual Studio or VS C++ Build Tools - do the following:](https://github.com/Gourieff/sd-webui-reactor#viii-for-windows-users-if-you-still-cannot-build-insightface-for-some-reasons-or-just-dont-want-to-install-visual-studio-or-vs-c-build-tools---do-the-following)
    Close (stop) your SD WebUI Server if it's running
    Download and put [prebuilt Insightface package](https://github.com/Gourieff/sd-webui-reactor/raw/main/example/insightface-0.7.3-cp310-cp310-win_amd64.whl) into the stable-diffusion-webui (or SD.Next) root folder (where you have "webui-user.bat" file)
    From stable-diffusion-webui (or SD.Next) root folder run CMD and .\venv\Scripts\activate
    Then update your PIP: python -m pip install -U pip
    Then install Insightface: pip install insightface-0.7.3-cp310-cp310-win_amd64.whl
    Enjoy!_

Then I tried Gourieff's Reactor (and the newer Reactor-Force) which, in my opinion works better.

But for someone who is not a programmer, not familiar with python etc. it took some time and learning to be able to sucessfully carry out the instructions above.

So, for beginner who really need lots of help, making it part of the package would be very helpful, especially since Fooocus does not let you add extensions : )

**Reply by barepixels** [reply]
Fooocus is base on ComfyUI... so it will need Face Swap Lab or Reactor for ComfyUI.  Am kinda busy right now to confirm is such available for ComfyUI.

**Reply by KINGLIFER** [reply]
The link above has info on comfy install for faces faceswap. What about ipadaptor plus face (does the same thing and perhaps better). Since you have controlnet would it be an inconvenience to add that feature? It would only need a way to import the face image

**Reply by barepixels** [reply]
ipadaptor plus face is only available for SD 1.5.  For now lllyasviel Fooocus is SDXL only

**Reply by KINGLIFER** [reply]
There is for SDXL so again do you think this can be implemented?

**Reply by barepixels** [reply]
Thanks for head up.  https://github.com/cubiq/ComfyUI_IPAdapter_plus  Just got done downloading 2 new models  
ip-adapter_sdxl.bin
ip-adapter-plus-face_sdxl_vit-h.bin
Going to play with it now in Comfy

**Comment by barepixels** [comment]
it works YAY.  Model used https://civitai.com/models/139562?modelVersionId=169921
![woman](https://github.com/lllyasviel/Fooocus/assets/92839033/50efc06e-e82e-4f64-9fe3-a70bb3460acc)
![IPAdapter_00057_](https://github.com/lllyasviel/Fooocus/assets/92839033/46c91a6b-0dea-43e1-8d5e-89866c60b1cf)
Thanks KINGLIFER

**Reply by KINGLIFER** [reply]
okay but will this be implemented with fooocus?

**Comment by KINGLIFER** [comment]
sigh

**Comment by lllyasviel** [comment]
in 790

**Comment by KitsonBroadhurst** [comment]
Got face swap working on Fooocus: https://www.youtube.com/watch?v=nWvOzBG1k_k

**Comment by machineminded** [comment]
I created a fork of Fooocus with an inswapper/insightface integration:

https://github.com/machineminded/Fooocus-inswapper#usage

**Reply by dav30383** [reply]
very good! Its possible to use this in Google Colab? i Love Fooocus, but the integrated faceswap is totally shit

**Reply by machineminded** [reply]
Probably - I'm not really familiar with Google Colab.  I will have to look into it but I would think it is possible.

**Comment by dav30383** [comment]
It would truly be a spectacle! For people like me who don't have a
high-performance video card, using Colab is perfect

Il lun 29 gen 2024, 15:58 machineminded ***@***.***> ha
scritto:

**Comment by Gracefruit** [comment]
Fooocus is really good for [face swap](https://www.deepswap.ai/), but Installation is difficult for beginners.
![pic](https://github.com/lllyasviel/Fooocus/assets/171776045/4214326d-a73b-48b2-aefd-d54a5e912cf1)
![pic(1)](https://github.com/lllyasviel/Fooocus/assets/171776045/f799bedd-d65c-4e2a-8d1e-5e8a4fb147d0)

**Comment by drdancm** [comment]
Gracefruit,

I use Stability Matrix, which makes Fooocus, and other package installations as simple as one mouse click. However, I have not had much luck using FaceSwap feature in the last couple of Fooocus updates to work really well. For me Reactor works far more easily with A1111 or with Forge.  Reactor does not require tuning lots of  parameters to get a good swap at all.  

If you would not mind giving some helpful configuration hints that would be great. I followed Broadhurst's youtube tutorial, but as I said, I could not get any swaps even half as good as with Reactor. Also, sadly machineminded fork with Reactor, has not been made available from Stability Matrix.

Thanks!

**Reply by machineminded** [reply]
@mashb1t Would you be open to integrating inswapper_128 into the main repo?

**Reply by mashb1t** [reply]
@machineminded would have to compare the methods first, but in general yes

**Reply by kalle07** [reply]
may take a look at 
sd-webui-reactor
https://github.com/Gourieff/sd-webui-reactor
and may copy and paste  ;)
